Change RemoteLayerTreeDrawingAreaProxy::ProcessState to track a set of pending commits, rather than a single state machine.

mattwoodrow opened the pull request:
#### a327fd79c613902d96e90b7cf8c43e92e6885835
<pre>
Change RemoteLayerTreeDrawingAreaProxy::ProcessState to track a set of pending commits, rather than a single state machine.
<a href="https://bugs.webkit.org/show_bug.cgi?id=303715">https://bugs.webkit.org/show_bug.cgi?id=303715</a>
&lt;<a href="https://rdar.apple.com/problem/166025253">rdar://problem/166025253</a>&gt;

Reviewed by Gerald Squelart.

Trying to track potentially multiple commits with a single state machine is
complicated as the set of potential states is large.

Replace the state variant with a vector of pending commits, so that each
potential state can be tracked per-commit.

Fixes a bug where the existing MESSAGE_CHECKs failed, where we thought the next
step was waiting for display did refresh, but another commit arrives.

This also enables the multiple frame setting by default so that EWS tests run
and expose bugs like the above.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Shared/MonotonicObjectIdentifier.h:
(WebKit::MonotonicObjectIdentifier::isHashTableEmptyValue const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::ProcessState::ProcessState):
(WTF::HashTraits&lt;WebKit::ProcessState&gt;::emptyValue):
(WTF::HashTraits&lt;WebKit::ProcessState&gt;::isEmptyValue):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::RemoteLayerTreeDrawingAreaProxy):
(WebKit::RemoteLayerTreeDrawingAreaProxy::addRemotePageDrawingAreaProxy):
(WebKit::ProcessState::ProcessState):
(WebKit::RemoteLayerTreeDrawingAreaProxy::nextMainFrameLayerTreeTransactionID const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::processStateForConnection):
(WebKit::RemoteLayerTreeDrawingAreaProxy::processStateForIdentifier const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::notifyPendingCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::maybePauseDisplayRefreshCallbacks):
(WebKit::operator&lt;&lt;):
(WebKit::ProcessState::canSendDisplayDidRefresh):
(WebKit::RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay):
(WebKit::RemoteLayerTreeDrawingAreaProxy::waitForDidUpdateActivityState):
(WebKit::RemoteLayerTreeDrawingAreaProxy::ProcessState::canSendDisplayDidRefresh):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::windowScreenDidChange):

Canonical link: <a href="https://commits.webkit.org/305784@main">https://commits.webkit.org/305784@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af3cba1f698245e116e17888e371797b6862efeb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/139121 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/11494 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/618 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/147248 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/92188 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/6d8f3207-122a-4a6f-885e-5cf2af244989) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/12201 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/11645 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/106503 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/77582 "5 flakes 5 failures") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/6d6eee37-c4a4-42c7-ba95-341bd432db54) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/142068 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/9216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/124611 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/87370 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/829c67ad-944d-45b4-adee-d69a97990e7f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/8766 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/6542 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/7540 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/131094 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/118217 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/150027 "Built successfully") | | 
| [⏳ 🛠 🧪 jsc-debug-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-O3-Debug-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/11179 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/521 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/114886 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/11192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/9457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/115198 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29321 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/9113 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/120960 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/66081 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/11221 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/488 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/170392 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/10957 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/74879 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/44367 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/11160 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/11009 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -52,6 +52,62 @@ class RemoteAnimationStack;
+// displayDidRefresh is sent to the WebProcess, and it responds
+// with a commitLayerTree message (ideally before the next
+// displayDidRefresh, otherwise we mark it as missed and send
+// it when commitLayerTree does arrive).
```
Is this the right location for this comment? It seems a bit disconnected from `ProcessState`.
I see you only moved things here, but maybe it's an opportunity to find a better place for this comment, and maybe reword it (missed -> delayed, etc.), if it's even still needed?

squelart commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -52,6 +52,62 @@ class RemoteAnimationStack;
+template<> struct HashTraits<WebKit::ProcessState> : SimpleClassHashTraits<WebKit::ProcessState> {
+    static constexpr bool emptyValueIsZero = HashTraits<WebKit::TransactionID>::emptyValueIsZero;
+    static WebKit::ProcessState emptyValue() { return { HashTraits<WebKit::TransactionID>::emptyValue() }; }
+    static bool isEmptyValue(const WebKit::ProcessState& value) { return !value.nextLayerTreeTransactionID.object(); }
```
Shouldn't you use `HashTraits<WebKit::TransactionID>::isEmptyValue()` to be consistent with `emptyValue()` above?

squelart commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -273,33 +283,27 @@ - (void)drawInContext:(CGContextRef)cgContext
+            didRefreshDisplay(&connection);
+        }
+    } else {
+        // This frame is still pending, it'll be sent when the WebProcess decides its ready.
```
its -> it's

squelart commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -754,39 +760,30 @@ - (void)drawInContext:(CGContextRef)cgContext
+        ASSERT(state.pendingCommits.size());
+        if (state.pendingCommits.last().delayState == CommitDelayState::Pending) {
+            state.pendingCommits.last().delayState = CommitDelayState::Delayed;
             LOG_WITH_STREAM(RemoteLayerTree, stream << "RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay stil waiting on commit, marked as missed");
```
stil -> still

squelart replied:
Should "marked as missed" be changed to "marked as delayed" to match the new state name?
And you may want to check all comments and log messages for this change.

squelart commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -273,33 +283,27 @@ - (void)drawInContext:(CGContextRef)cgContext
+        }
+    } else {
+        // This frame is still pending, it'll be sent when the WebProcess decides its ready.
+        // Use the IntentionallyDeferred state so that we don't think its late
```
its -> it's

squelart commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -841,17 +838,16 @@ - (void)drawInContext:(CGContextRef)cgContext
+        if (!m_webPageProxyProcessState.pendingCommits.size())
+            error = didRefreshDisplay(m_webPageProxyProcessState, connection.get());
+        else {
+            // Only the first outstanding frame can be in NotifyPendingCommitLayerTree state
```
"First" may be a bit ambiguous: It's the first item in the vector, but it's the most recent frame, right?
So I'd suggest "most recent" or "latest".
(If I'm incorrect, I'd need to re-read everything!)

mattwoodrow replied:
You are absolutely correct :)

squelart reviewed (approved):
Looks good to me (only minor mostly-non-code nits), but I haven't thought about every possible scenario.
Please let me know if I should spend more time, or get a 2nd opinion for more certainty.

mattwoodrow committed:
Change RemoteLayerTreeDrawingAreaProxy::ProcessState to track a set of pending commits, rather than a single state machine.
https://bugs.webkit.org/show_bug.cgi?id=303715
<rdar://problem/166025253>

Reviewed by Gerald Squelart.

Trying to track potentially multiple commits with a single state machine is
complicated as the set of potential states is large.

Replace the state variant with a vector of pending commits, so that each
potential state can be tracked per-commit.

Fixes a bug where the existing MESSAGE_CHECKs failed, where we thought the next
step was waiting for display did refresh, but another commit arrives.

This also enables the multiple frame setting by default so that EWS tests run
and expose bugs like the above.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Shared/MonotonicObjectIdentifier.h:
(WebKit::MonotonicObjectIdentifier::isHashTableEmptyValue const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::ProcessState::ProcessState):
(WTF::HashTraits<WebKit::ProcessState>::emptyValue):
(WTF::HashTraits<WebKit::ProcessState>::isEmptyValue):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::RemoteLayerTreeDrawingAreaProxy):
(WebKit::RemoteLayerTreeDrawingAreaProxy::addRemotePageDrawingAreaProxy):
(WebKit::ProcessState::ProcessState):
(WebKit::RemoteLayerTreeDrawingAreaProxy::nextMainFrameLayerTreeTransactionID const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::processStateForConnection):
(WebKit::RemoteLayerTreeDrawingAreaProxy::processStateForIdentifier const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::notifyPendingCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::maybePauseDisplayRefreshCallbacks):
(WebKit::operator<<):
(WebKit::ProcessState::canSendDisplayDidRefresh):
(WebKit::RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay):
(WebKit::RemoteLayerTreeDrawingAreaProxy::waitForDidUpdateActivityState):
(WebKit::RemoteLayerTreeDrawingAreaProxy::ProcessState::canSendDisplayDidRefresh):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::windowScreenDidChange):

Canonical link: https://commits.webkit.org/305784@main
